servers/republik/modules/crowdfundings/lib/activateYearlyMembership.js

```diff
@@ -0,0 +1,80 @@
+const moment = require('moment')
+const { ascending } = require('d3-array')
+
+module.exports = async (memberships, pgdb, dryRun) => {
+  const membershipTypeIndex = (
+    await pgdb.public.membershipTypes.findAll()
+  ).reduce(
+    (index, type) => {
+      index[type.id] = type
+      return index
+    },
+    {}
+  )
+  const packagesIndex = (
+    await pgdb.public.packages.findAll()
+  ).reduce(
+    (index, p) => {
+      index[p.id] = p
+      return index
+    },
+    {}
+  )
+
+  const yearlyMemberships = memberships.filter(m => membershipTypeIndex[m.membershipTypeId].name !== 'MONTHLY_ABO')
+  if (!yearlyMemberships.length) {
+    return
+  }
+
+  const membershipPeriods = await pgdb.public.membershipPeriods.find({
+    membershipId: memberships.map(m => m.id)
+  })
+  const unusedMemberships = yearlyMemberships.filter(m => !membershipPeriods.find(p => p.membershipId === m.id))
+  if (!unusedMemberships.length) {
+    return
+  }
+
+  const pledges = await pgdb.public.pledges.find({
+    id: memberships.map(m => m.pledgeId)
+  })
+
+  const inactiveMemberships = unusedMemberships.filter(m => {
+    const pledge = pledges.find(p => p.id === m.pledgeId)
+    return packagesIndex[pledge.packageId].name !== 'ABO_GIVE' || !m.voucherCode
+  })
+  if (!inactiveMemberships.length) {
+    return
+  }
+
+  inactiveMemberships.sort((a, b) => ascending(a.sequenceNumber, b.sequenceNumber))
+
+  const electedMembership = (
+    inactiveMemberships.find(m => m.voucherCode === null) ||
```

Comment:
Is this line necessary considering `|| !m.voucherCode` in line 43?

Reply:
yes because above it ensure that only none-give abos or redeemed give abos are considere for activation. Here it says that a reedemed abo should be used first (same logic as January activation).

Reply:
Ah, so basically a `null` voucherCode represents redeemed, while a "normal" membership would have an `undefined` voucherCode, correct? (Just trying to understand)

Reply:
normal inactive memberships have a voucherCode string